-1 
Tested at: da99a5e76ae262201078cc44e8f99adef33f76e5
 I found an error when building: 
 
 <pre>   inline void throwMissingLabel(const std::string& what, const std::string& bad_label, 
               ^
In file included from /tmp/cmsbuild/workspace/ib-any-integration/CMSSW_7_6_X_2015-10-03-2300/src/DataFormats/PatCandidates/interface/Photon.h:20:0,
                 from /tmp/cmsbuild/workspace/ib-any-integration/CMSSW_7_6_X_2015-10-03-2300/src/RecoHI/HiEgammaAlgos/test/patTest/PATHIPhotonTestModule.cc:36:
/tmp/cmsbuild/workspace/ib-any-integration/CMSSW_7_6_X_2015-10-03-2300/src/DataFormats/PatCandidates/interface/PATObject.h: In member function 'float pat::PATObject<ObjectType>::userFloat(const string&) const [with ObjectType = reco::Photon; std::string = std::basic_string<char>]':
/tmp/cmsbuild/workspace/ib-any-integration/CMSSW_7_6_X_2015-10-03-2300/src/DataFormats/PatCandidates/interface/PATObject.h:774:3: error: control reaches end of non-void function [-Werror=return-type]
   }
   ^
cc1plus: some warnings being treated as errors
gmake: *** [tmp/slc6_amd64_gcc493/src/RecoHI/HiEgammaAlgos/test/RecoHIHiEgammaAlgos_testAnalyzers/patTest/PATHIPhotonTestModule.o] Error 1
>> Compiling  /tmp/cmsbuild/workspace/ib-any-integration/CMSSW_7_6_X_2015-10-03-2300/src/RecoHI/HiEgammaAlgos/src/EcalClusterIsoCalculator.cc 
</pre> 
 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11574/8598/summary.html

--- a/DataFormats/PatCandidates/interface/PATObject.h
+++ b/DataFormats/PatCandidates/interface/PATObject.h
@@ -771,6 +771,7 @@ namespace pat {
       return userFloats_[std::distance(userFloatLabels_.cbegin(),it)];
     }
     throwMissingLabel("UserFloat",key,userFloatLabels_);
+    return std::numeric_limits<float>::quiet_NaN();
   }
 
   template <class ObjectType>
@@ -801,6 +802,7 @@ namespace pat {
       return userInts_[std::distance(userIntLabels_.cbegin(),it)];
     }
     throwMissingLabel("UserInt",key,userIntLabels_);
+    return std::numeric_limits<int>::max();
   }
 
   template <class ObjectType>

--- a/DataFormats/PatCandidates/interface/PATObject.h
+++ b/DataFormats/PatCandidates/interface/PATObject.h
@@ -319,8 +319,10 @@ namespace pat {
       }
 
       /// Get user-defined float
-      /// Note: it will return NaN if the key is not found; you can check if the key exists with 'hasUserFloat' method.
+      /// Note: throws if the key is not found; you can check if the key exists with 'hasUserFloat' method.
       float userFloat( const std::string & key ) const;
+      /// return a range of values corresponding to key
+      std::vector<float> userFloatRange( const std::string& key ) const;
       /// a CINT-friendly interface
       float userFloat( const char* key ) const { return userFloat( std::string(key) ); }
       
@@ -337,10 +339,12 @@ namespace pat {
       bool hasUserFloat( const char* key ) const {return hasUserFloat( std::string(key) );}
 
       /// Get user-defined int
-      /// Note: it will return 0 if the key is not found; you can check if the key exists with 'hasUserInt' method.
+      /// Note: throws if the key is not found; you can check if the key exists with 'hasUserInt' method.
       int32_t userInt( const std::string & key ) const;
+      /// returns a range of values corresponding to key
+      std::vector<int> userIntRange( const std::string& key ) const;
       /// Set user-defined int
-      void addUserInt( const std::string & label,  int32_t data, const bool overwrite = false );
+      void addUserInt( const std::string & label,  int32_t data );
       /// Get list of user-defined int names
       const std::vector<std::string> & userIntNames() const  { return userIntLabels_; }
       /// Return true if there is a user-defined int with a given name
@@ -774,6 +778,17 @@ namespace pat {
     return std::numeric_limits<float>::quiet_NaN();
   }
 
+  template<class ObjectType>
+  std::vector<float> PATObject<ObjectType>::userFloatRange( const std::string& key ) const {
+    auto range = std::equal_range(userFloatLabels_.cbegin(),userFloatLabels_.cend(),key);
+    std::vector<float> result;
+    result.reserve(std::distance(range.first,range.second));
+    for( auto it = range.first; it != range.second; ++it ) {
+      result.push_back(userFloats_[std::distance(userFloatLabels_.cbegin(),it)]);
+    }
+    return result;
+  }
+
   template <class ObjectType>
   void PATObject<ObjectType>::addUserFloat( const std::string & label,
 					    float data,
@@ -784,12 +799,9 @@ namespace pat {
     if( it == userFloatLabels_.end() || *it != label ) {      
       userFloatLabels_.insert(it,label);
       userFloats_.insert(userFloats_.begin()+dist,data);
-    } else if( overwrite && *it == label  ) {
-      userFloats_[dist] = data;
-    } else {
-      edm::LogWarning("addUserFloat") 
-        << "Attempting to add userFloat: " 
-        << label << " = " << data << " failed, value exists already!";
+    } else if( *it == label  ) {
+      //create a range by adding behind the first entry
+      userFloats_.insert(userFloats_.begin()+dist+1, data); 
     }
   }
 
@@ -805,22 +817,29 @@ namespace pat {
     return std::numeric_limits<int>::max();
   }
 
+  template<class ObjectType>
+  std::vector<int> PATObject<ObjectType>::userIntRange( const std::string& key ) const{
+    auto range = std::equal_range(userIntLabels_.cbegin(),userIntLabels_.cend(),key);
+    std::vector<int> result;
+    result.reserve(std::distance(range.first,range.second));
+    for( auto it = range.first; it != range.second; ++it ) {
+      result.push_back(userInts_[std::distance(userIntLabels_.cbegin(),it)]);
+    }
+    return result;
+  }
+
   template <class ObjectType>
   void PATObject<ObjectType>::addUserInt( const std::string &label,
-                                          int data,
-                                          const bool overwrite )
+                                          int data )
   {
     auto it = std::lower_bound(userIntLabels_.begin(),userIntLabels_.end(),label);
     const auto dist = std::distance(userIntLabels_.begin(),it);
     if( it == userIntLabels_.end() || *it != label ) { 
       userIntLabels_.insert(it,label);
       userInts_.insert(userInts_.begin()+dist,data);
-    } else if( overwrite && *it == label ) {
-      userInts_[dist] = data;
-    } else {
-      edm::LogWarning("addUserInt") 
-        << "Attempting to add userInt: " 
-        << label << " = " << data << " failed, value exists already!";
+    } else if( *it == label ) {
+      //create a range by adding behind the first entry
+      userInts_.insert(userInts_.begin()+dist+1,data);
     }
   }
 

--- a/DataFormats/PatCandidates/interface/throwMissingLabel.h
+++ b/DataFormats/PatCandidates/interface/throwMissingLabel.h
@@ -5,8 +5,8 @@
 #include <vector>
 
 namespace pat {
-  inline void throwMissingLabel(const std::string& what, const std::string& bad_label, 
-                                const std::vector<std::string>& available);
+  void throwMissingLabel(const std::string& what, const std::string& bad_label, 
+                         const std::vector<std::string>& available);
 }
 
 #endif

